In [1]:
import numpy as np
import pandas as pd
from sklearn.decomposition import PCA
from sklearn import preprocessing

import matplotlib.pylab as plt

In [5]:
df = pd.read_csv('RidingMowers.csv')

In [6]:
df.head()

Income  Lot_Size Ownership
0    60.0      18.4     Owner
1    85.5      16.8     Owner
2    64.8      21.6     Owner
3    61.5      20.8     Owner
4    87.0      23.6     Owner

In [10]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis


In [11]:
da_reg = LinearDiscriminantAnalysis()

In [12]:
da_reg.fit(df.drop(columns=['Ownership']), df['Ownership'])

LinearDiscriminantAnalysis()

In [18]:
df.head()

Income  Lot_Size Ownership
0    60.0      18.4     Owner
1    85.5      16.8     Owner
2    64.8      21.6     Owner
3    61.5      20.8     Owner
4    87.0      23.6     Owner

In [13]:
print('Coefficients', da_reg.coef_)

Coefficients [[0.1002303  0.78518471]]


In [14]:
print('Intercept', da_reg.intercept_)

Intercept [-21.73876167]


For instance, the first household has an income of $60K and a lot size of 18.4K ft
(0.1)(60) + (0.79)(18.4) − 21.74 = −1.28. 
Since the value is negative,
the household is (mis)classified by the model as a nonowner.

In [16]:
X_new = [[60, 18.4]]
# Get predicted class
predicted_class = da_reg.predict(X_new)
print(predicted_class)

['Nonowner']


/home/bardend/Documents/Cursos/8/Data_Mining/venv/lib/python3.10/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearDiscriminantAnalysis was fitted with feature names
  warnings.warn(


### Classification with probability


In [20]:
da_reg = LinearDiscriminantAnalysis()
da_reg.fit(mower_df.drop(columns=['Ownership']), mower_df['Ownership'])
result_df = mower_df.copy()
result_df['Dec. Function'] = da_reg.decision_function(mower_df.drop(columns=['Ownership']))
result_df['Prediction'] = da_reg.predict(mower_df.drop(columns=['Ownership']))
result_df['p(Owner)'] = da_reg.predict_proba(mower_df.drop(columns=['Ownership']))[:, 1]
result_df.head()

Income  Lot_Size Ownership  Dec. Function Prediction  p(Owner)
0    60.0      18.4     Owner      -1.277545   Nonowner  0.217968
1    85.5      16.8     Owner       0.022032      Owner  0.505508
2    64.8      21.6     Owner       1.716152      Owner  0.847632
3    61.5      20.8     Owner       0.757244      Owner  0.680755
4    87.0      23.6     Owner       5.511634      Owner  0.995977